capsules/extra/src/hc_sr04.rs

```diff
@@ -0,0 +1,158 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+use core::cell::Cell;
+
+use kernel::hil::gpio::{Client, InterruptEdge};
+use kernel::hil::sensors::{self, Distance, DistanceClient};
+use kernel::hil::time::{Alarm, Ticks};
+use kernel::hil::time::{AlarmClient, ConvertTicks};
+use kernel::utilities::cells::OptionalCell;
+use kernel::ErrorCode;
+
+/// Maximum duration for the echo pulse to be measured in milliseconds.
+// As specified in the datasheet:
+// https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+// the maximum time for the echo pulse to return is around 38 milliseconds
+// for a maximum distance of approximately 4 meters. We use a slightly
+// higher value to account for possible variations in measurement.
+pub const MAX_DISTANCE_ECHO: u32 = 50;
+
+/// Speed of sound in air in mm/s.
+// The speed of sound is approximately 343 meters per second, which
+// translates to 343,000 millimeters per second. This value is used
+// to calculate the distance based on the time it takes for the echo
+// to return.
+pub const SPEED_OF_SOUND: u32 = 343000;
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Distance as usize;
+
+#[derive(Copy, Clone, PartialEq)]
+/// Status of the sensor.
+pub enum Status {
+    Idle,         // Sensor is idle.
+    TriggerPulse, // Sending ultrasonic pulse.
+    EchoStart,    // Interrupt on the rising edge.
+    EchoEnd,      // Interrupt on the falling edge.
+}
+
+/// HC-SR04 Ultrasonic Distance Sensor Driver
+pub struct HcSr04<'a, A: Alarm<'a>> {
+    trig: &'a dyn kernel::hil::gpio::Pin,
+    echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+    alarm: &'a A,
+    start_time: Cell<u32>,
+    end_time: Cell<u32>,
```

Comment:
Why do you need an `end_time` member of the struct?

Reply:
I've added a timeout in the `Status::EchoStart` case so that if we don't get an echo within 50ms, the measurement will fail and made sure the timer event gets canceled if we get a GPIO interrupt.
I've also put in a check to make sure the echo pulse duration doesn't exceed the maximum expected time. If it does, we throw an error `ErrorCode::INVAL` since that's likely a bad reading.